addons/source-python/packages/source-python/memory/manager.py

```diff
@@ -646,34 +659,48 @@ def function(
         if return_type not in DataType.values:
             return_type = self.create_converter(return_type)
 
+        # Store the function cache
+        func = None
+
         class fget(object):
+            def __set_name__(fget_self, owner, name):
+                fget_self.name = name
+
             def __get__(fget_self, obj, cls=None):
+                nonlocal func
                 if cls is None:
                     if obj is None:
                         return fget_self
                     else:
                         cls = obj.__class__
 
-                if cls._binary is None:
-                    raise ValueError('_binary was not specified.')
+                if func is None:
+                    if cls._binary is None:
+                        raise ValueError('_binary was not specified.')
 
-                # Create a binary object
-                binary = find_binary(cls._binary, cls._srv_check)
+                    # Create a binary object
+                    binary = find_binary(cls._binary, cls._srv_check)
 
-                # Create the function object
-                func = binary[identifier].make_function(
-                    convention,
-                    args,
-                    return_type
-                )
+                    # Create the function object and cache it
+                    func = binary[identifier].make_function(
+                        convention,
+                        args,
+                        return_type
+                    )
+                    func.__doc__ = doc
 
                 # Called with a this pointer?
                 if obj is not None:
-                    # Wrap the function using MemberFunction, so we don't have
-                    # to pass the this pointer anymore
-                    func = MemberFunction(self, return_type, func, obj)
+                    # Wrap the function using MemberFunction,
+                    # so we don't have to pass the this pointer anymore
+                    m_func = MemberFunction(self, return_type, func, obj)
+                    m_func.__doc__ = doc
+
+                    # Set the MemberFunction as an attribute to the instance.
+                    setattr(obj, fget_self.name, m_func)
```

Comment:
So there is no way to set MemberFunction to the instance.
Can't we just use types.MethodType, for example, and set the method to the instance with a name like _function_name, with the exception for normal function call?

Reply:
OK, that doesn't really make much sense. Forget about it.

Reply:
This was discarded. Here are the new test results.
Output :
```
TypeManager.function Test
4.355441149324179 Test._spawn
4.830625455826521 test._spawn
4.801660872995853 Test()._spawn
↓
0.023133378475904465 Test._spawn
0.29126785323023796 test._spawn
0.293557733297348 Test()._spaw

TypeManager.virtual_function Test
0.5497394129633904 test.blind
0.5683586820960045 make_object(Test, pointer).blind
↓
0.46108797565102577 test.blind
0.4727563001215458 make_object(Test, pointer).blind
```

Reply:
> Unless I completely misunderstand what you mean, the result would be exactly the same because `self` would then be held by `MethodType.__self__` instead of `MemberFunction._this`.

You're right, MethodType needed a weakref.proxy, the test I was doing was `setattr(CustomType, "function_name", lambda self, *args: CustomType.function_name(self, *args))`, I thought MethodType was the same thing, but I was wrong. But in the end, with lambda function, call_trampoline and skip_hooks cannot be used.

Are there any other issues with this pull request?

Reply:
> I haven't tested, but based on the code, I can tell you that the `_ptr()` call on [L628](https://github.com/Source-Python-Dev-Team/Source.Python/pull/479/files#diff-9c8c742848f8ac84b665979c76e7d1a7d4bc4b825b81e443835be54ab109196bR628) is redundant since `ptr` is already a `Pointer`.

Done.

> That said, I think `virtual_function` could probably make great use of `CachedProperty`.

Using `CachedProperty` leaks `ptr/_this`. Is there a way to use `CachedProperty` without causing leaks?

Reply:
Yes, it does not work regardless of the value of unbound.